File: src/main/java/dev/rvbsm/fsit/api/event/PassedUseEntityCallback.java
package dev.rvbsm.fsit.api.event;

import net.minecraft.entity.Entity;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.util.ActionResult;
import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;

import org.jetbrains.annotations.NotNull;

@FunctionalInterface
public interface PassedUseEntityCallback {

    Event<PassedUseEntityCallback> EVENT = EventFactory.createArrayBacked(
        PassedUseEntityCallback.class, (listeners) -> (player, world, entity) -> {
            for (PassedUseEntityCallback listener : listeners) {
                final ActionResult result = listener.interact(player, world, entity);

                if (result != ActionResult.PASS) {
                    return result;
                }
            }

            return ActionResult.PASS;
        });

    ActionResult interact(@NotNull ServerPlayerEntity player, @NotNull ServerWorld world, @NotNull Entity entity);
}

File: src/main/kotlin/dev/rvbsm/fsit/event/PassedUseEntityCallback.kt
package dev.rvbsm.fsit.event

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.entity.Entity
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.server.world.ServerWorld
import net.minecraft.util.ActionResult

fun interface PassedUseEntityCallback {
    fun interactEntity(player: ServerPlayerEntity, world: ServerWorld, entity: Entity): ActionResult

    companion object {
        @JvmField
        val EVENT: Event<PassedUseEntityCallback> =
            EventFactory.createArrayBacked(PassedUseEntityCallback::class.java) { listeners ->
                PassedUseEntityCallback { player, world, entity ->
                    for (listener in listeners) {
                        val result = listener.interactEntity(player, world, entity)

                        if (result != ActionResult.PASS) {
                            return@PassedUseEntityCallback result
                        }
                    }

                    return@PassedUseEntityCallback ActionResult.PASS
                }
            }
    }
}
